<!--
  ~ Copyright 2015 OpenCB
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~     http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->

<link rel="import" href="../opencga-active-filters.html">
<link rel="import" href="variant-browser-grid.html">
<link rel="import" href="../../cellbase/variation/cellbase-variantannotation-view.html">
<link rel="import" href="variant-beacon-network.html">
<link rel="import" href="variant-genome-browser.html">

<dom-module id="variant-browser-view">
    <template>
        <style include="jso-styles">
            .browser-variant-tab-title {
                font-size: 115%;
                font-weight: bold;
            }

            th {
                text-align: center
            }

            .icon-padding {
                padding-left: 4px;
                padding-right: 8px;
            }
        </style>

        <!--<div>-->
        <opencga-active-filters opencga-client="{{opencgaClient}}" query="[[query]]" filters="{{config.filters}}"
                                filter-bioformat="VARIANT" default-study="{{study.alias}}" alias="{{activeFilterAlias}}" refresh="[[search]]">
        </opencga-active-filters>

        <div class="col-md-12" style="padding: 5px 0px 5px 0px">
            <div class="btn-toolbar" role="toolbar" aria-label="..." style="padding: 10px 0px;margin-left: 0px">
                <div id="{{prefix}}LeftToolbar" style="padding-bottom: 0px">
                    <div class="btn-group" role="group">
                        <button type="button" class="btn btn-success variant-browser-view-buttons active" data-view="TableResult" on-click="_changeView">
                            <i class="fa fa-table icon-padding" aria-hidden="true" data-view="TableResult" on-click="_changeView"></i> Table Result
                        </button>
                        <button type="button" class="btn btn-success variant-browser-view-buttons" data-view="Summary" on-click="_changeView">
                            <i class="fa fa-line-chart icon-padding" aria-hidden="true" data-view="Summary" on-click="_changeView"></i> Summary (Beta)
                        </button>
                        <button type="button" class="btn btn-success variant-browser-view-buttons" data-view="GenomeBrowser" on-click="_changeView">
                            <i class="fa fa-list-alt icon-padding" aria-hidden="true" data-view="GenomeBrowser" on-click="_changeView"></i> Genome Browser (Beta)
                        </button>
                    </div>
                </div>
            </div>
        </div>

        <div>
            <!-- First TAB -->
            <div id="{{prefix}}TableResult" class="variant-browser-content">
                <!-- Variant Browser Grid -->
                <variant-browser-grid study="{{study}}" project="{{project}}" cohorts="{{cohorts}}"
                                      opencga-client="{{opencgaClient}}" cellbase-client="{{cellbaseClient}}"
                                      population-frequencies="{{populationFrequencies}}" active="{{active}}"
                                      protein-substitution-scores="{{proteinSubstitutionScores}}"
                                      consequence-types="{{consequenceTypes}}" query="[[query]]" search="{{search}}"
                                      variant="{{variant}}" on-selected="selectedGene"
                                      on-selectvariant="onSelectVariant" style="font-size: 12px" config="{{config}}"
                                      summary="{{summary}}" samples="{{samples}}">
                </variant-browser-grid>


                <!-- Bottom tabs with specific variant information -->
                <div style="padding-top: 20px" hidden$="{{!checkVariant(variant)}}">
                    <h3>Variant: {{variant}}</h3>
                    <div style="padding-top: 20px">
                        <ul id="{{prefix}}ViewTabs" class="nav nav-tabs" role="tablist">
                            <li role="presentation" class="active"><a href="#{{prefix}}Annotation" role="tab"
                                                                      data-toggle="tab"
                                                                      class="browser-variant-tab-title">Advanced
                                Annotation</a></li>
                            <li role="presentation"><a href="#{{prefix}}Genotype" role="tab" data-toggle="tab"
                                                       class="browser-variant-tab-title">Genotype Stats</a></li>
                            <!--<li role="presentation"><a href="#{{prefix}}GenomeView" role="tab" data-toggle="tab" class="browser-variant-tab-title">Genome Context</a></li>-->
                            <li role="presentation"><a href="#{{prefix}}Beacon" role="tab" data-toggle="tab"
                                                       class="browser-variant-tab-title">Beacon Network</a></li>
                        </ul>

                        <div class="tab-content" style="height: 680px">
                            <!--Annotation Tab-->
                            <div role="tabpanel" class="tab-pane active" id="{{prefix}}Annotation">
                                <cellbase-variantannotation-view data="{{variant}}" prefix="{{prefix}}"
                                                                 cellbase-client="{{cellbaseClient}}"
                                                                 mode="vertical"
                                                                 hash-fragment-credentials="{{hashFragmentCredentials}}"
                                                                 consequence-types="{{consequenceTypes}}"
                                                                 protein-substitution-scores="{{proteinSubstitutionScores}}"
                                                                 style="font-size: 12px">
                                </cellbase-variantannotation-view>
                            </div>

                            <!--Genotypes & Files Tab-->
                            <div role="tabpanel" class="tab-pane" id="{{prefix}}Genotype">
                                <template is="dom-repeat" items="{{variantStudies}}" as="study" id="variantStudies">
                                    <div class="col-md-10" style="padding-top: 8px">
                                        <h4 style="font-weight: bold;padding: 20px 0px 0px 0px">
                                            <i class="fa fa-minus-circle" on-click="handleCollapseAction"
                                               data-id="{{getStudy(study.studyId)}}" style="cursor: pointer;"></i>&nbsp;{{getStudy(study.studyId)}}
                                        </h4>

                                        <div id="{{getStudy(study.studyId)}}">
                                            <div hidden$="{{!checkStatsAvailable(study.stats)}}">
                                                <table class="table table-bordered" style="width: 70%">
                                                    <thead style="background-color: #eee;">
                                                    <tr>
                                                        <th scope="col" rowspan="2">Cohort</th>
                                                        <th scope="col" rowspan="2">Reference</th>
                                                        <th scope="col" rowspan="2">Alternate</th>
                                                        <th scope="col" rowspan="2">MAF (allele)</th>
                                                        <th colspan="2" scope="colgroup">Allele Frequencies</th>
                                                        <th colspan="3" scope="colgroup">Genotype Frequencies</th>
                                                    </tr>
                                                    <tr>
                                                        <th scope="col">Reference (count)</th>
                                                        <th scope="col">Alternate (count)</th>
                                                        <th scope="col">0/0 (count)</th>
                                                        <th scope="col">0/1 (count)</th>
                                                        <th scope="col">1/1 (count)</th>
                                                    </tr>
                                                    </thead>
                                                    <tbody>
                                                    <template is="dom-repeat" items="{{study.stats}}" as="cohort">
                                                        <tr>
                                                            <td>{{cohort.name}}</td>
                                                            <td>{{cohort.value.refAllele}}</td>
                                                            <td>{{cohort.value.altAllele}}</td>
                                                            <td>{{cohort.value.maf}} ({{cohort.value.mafAllele}})
                                                            </td>
                                                            <td>{{cohort.value.refAlleleFreq}}
                                                                ({{cohort.value.refAlleleCount}})
                                                            </td>
                                                            <td>{{cohort.value.altAlleleFreq}}
                                                                ({{cohort.value.altAlleleCount}})
                                                            </td>
                                                            <td>{{cohort.value.genotypesFreq.homref}}
                                                                ({{cohort.value.genotypesCount.homref}})
                                                            </td>
                                                            <td>{{cohort.value.genotypesFreq.het}}
                                                                ({{cohort.value.genotypesCount.het}})
                                                            </td>
                                                            <td>{{cohort.value.genotypesFreq.homalt}}
                                                                ({{cohort.value.genotypesCount.homalt}})
                                                            </td>
                                                        </tr>
                                                    </template>
                                                    </tbody>
                                                </table>
                                            </div>
                                            <div hidden$="{{checkStatsAvailable(study.stats)}}">
                                                No Genotypes statistics available
                                            </div>
                                        </div>
                                        <br>
                                    </div>
                                </template>
                            </div>

                            <!--Beacon network-->
                            <div role="tabpanel" class="tab-pane" id="{{prefix}}Beacon">
                                <br>
                                <button class="btn btn-primary" type="button" on-click="triggerBeacon">Search Beacon
                                    Network
                                </button>
                                <a data-toggle="tooltip"
                                   title="Beacon Network is a search engine across the world's public beacons. You can find it here - https://beacon-network.org/#/"><i
                                        class="fa fa-info-circle" aria-hidden="true"></i></a>
                                <br><br>
                                <variant-beacon-network hosts="{{beaconHosts}}" clear="{{variant}}"
                                                        variant="{{variantToBeacon}}"></variant-beacon-network>
                            </div>
                        </div>
                    </div>
                </div>
            </div>

            <!-- Second TAB -->
            <div id="{{prefix}}Summary" class="variant-browser-content" style="display: none">
                <br>
                <div class="col-md-6">
                    <h4>Consequence Type</h4>
                    <div id="{{prefix}}soAcc"></div>
                </div>
                <div class="col-md-6">
                    <h4>Gene</h4>
                    <div id="{{prefix}}genes"></div>
                </div>
                <div class="col-md-6">
                    <h4>Variant Type</h4>
                    <div id="{{prefix}}type"></div>
                </div>
                <div class="col-md-6">
                    <h4>Studies</h4>
                    <div id="{{prefix}}studies"></div>
                </div>
            </div>

            <!-- Third TAB -->
            <div id="{{prefix}}GenomeBrowser" class="variant-browser-content" style="display: none">
                <br>
                <br>
                <variant-genome-browser project="{{project}}" study="{{study}}" opencga-client="{{opencgaClient}}"
                                        cellbase-client="{{cellbaseClient}}"
                                        query={{query}} search={{search}} species={{config.species}}
                                        region="{{region}}" active="{{_genomeBrowserActive}}" full-screen="{{fullScreen}}">
                </variant-genome-browser>
            </div>

        </div>
        <!--</div>-->

        <div class="modal fade" id="{{prefix}}LoadingModal" data-backdrop="static" data-keyboard="false" tabindex="-1"
             role="dialog"
             aria-hidden="true" style="padding-top:15%; overflow-y:visible;">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <h3>Loading...</h3>
                    </div>
                    <div class="modal-body">
                        <div class="progress progress-striped active">
                            <div class="progress-bar progress-bar-success" style="width: 100%"></div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </template>

    <script>
        class VariantBrowserView extends Polymer.Element {

            constructor() {
                super();

                this.prefix = "varBrowserView" + Utils.randomString(6);
                this.summary = true;
            }

            static get is() {
                return 'variant-browser-view';
            }

            static get properties() {
                return {
                    title: {
                        type: String
                    },
                    project: {
                        type: Object
                    },
                    study: {
                        type: Object
                    },
                    samples: {
                        type: Array
                    },
                    cohorts: {
                        type: Object
                    },
                    data: {
                        type: Array,
                        value: []
                    },
                    opencgaClient: {
                        type: Object
                    },
                    cellbaseClient: {
                        type: Object
                    },
                    populationFrequencies: {
                        type: Object
                    },
                    proteinSubstitutionScores: {
                        type: Object
                    },
                    consequenceTypes: {
                        type: Object
                    },
                    beaconHosts: {
                        type: Array
                    },
                    query: {
                        type: Object
//                        notify: true,
                    },
                    search: {
                        type: Object,
                    },
                    variant: {
                        type: String,
                        value: "No variant selected",
                        observer: 'variantChanged'
                    },
                    variantStudies: {
                        type: Array,
                        value: [],
                        observer: "variantStudiesChanged"
                    },
                    summary: {
                        type: Boolean
                    },
                    fullScreen: {
                        type: Boolean
                    },
                    config: {
                        type: Object
                    }
                }
            }

            static get observers() {
                return [
                    'projectStudyObtained(project, study)'
                ]
            }

            _attachDom(dom) {
                this.appendChild(dom);
            }

            ready() {
                super.ready();

                this.genotypeColor = {
                    "0/0": "#6698FF",
                    "0/1": "#FFA500",
                    "1/0": "#FFA500",
                    "1/1": "#FF0000",
                    "./.": "#000000",
                    "0|0": "#6698FF",
                    "0|1": "#FFA500",
                    "1|0": "#FFA500",
                    "1|1": "#FF0000",
                    ".|.": "#000000",
                };

                this.activeFilterAlias = {
                    "annot-xref": "XRef",
                    "annot-biotype": "Biotype",
                    "annot-ct": "Consequence Types",
                    "alternate_frequency": "Population ALT Frequency",
                    "annot-functional-score": "CADD",
                    "protein_substitution": "Protein Substitution",
                    "annot-go": "GO",
                    "annot-hpo": "HPO"
                };

            }

            projectStudyObtained(project, study) {
                if (typeof this.project !== "undefined" && this.project.alias !== "" && typeof this.study !== "undefined" && this.study.alias !== "") {
                    this.hashFragmentCredentials = {
                        project: this.project.alias,
                        study: this.study.alias
                    };
                }
            }

            checkVariant(variant) {
                return variant.split(':').length > 2;
            }

            onSelectVariant(e) {
                this.variant = e.detail.id;
//                this.V = e.detail.variant;
//                this.display3D();
            }

            connectedCallback(){
                super.connectedCallback();
            }

            selectedGene(e) {
                this.dispatchEvent(new CustomEvent('propagate', {gene: e.detail.gene}));
            }

            checkStatsAvailable(stats) {
                return stats.length > 0;
            }

            variantChanged(e) {
                if (this.variant.split(':').length > 2) {
                    let [chromosome, start, ref, alt] = this.variant.split(":");
                    this.region = new Region(chromosome + ":" + start);
                    let params = {
                        ids: this.variant,
                        studies: this.project.alias + ":" + this.study.alias,
                        exclude: "annotation,studies.files,studies.samplesData"
                    };
                    let _this = this;
                    this.opencgaClient.variants().query(params)
                        .then(function (response) {
                            if (typeof response.response[0].result[0] !== "undefined") {
                                let _variantStudies = response.response[0].result[0].studies;
                                for (let i = 0; i < _variantStudies.length; i++) {
                                    let study = _variantStudies[i].studyId.split(':')[1];
                                    let statsObject = _variantStudies[i].stats;
                                    let statsArray = [];
                                    Object.keys(statsObject).map(key => {
                                        statsObject[key].maf = _this._freqFormatter(statsObject[key].maf || 0);
                                        statsObject[key].mafAllele = statsObject[key].mafAllele || '-';

                                        statsObject[key].refAlleleFreq = _this._freqFormatter(statsObject[key].refAlleleFreq || 0);
                                        statsObject[key].altAlleleFreq = _this._freqFormatter(statsObject[key].altAlleleFreq || 0);

                                        statsObject[key].genotypesFreq.homref = _this._freqFormatter(statsObject[key].genotypesFreq["0/0"] || statsObject[key].genotypesFreq["0|0"] || 0);
                                        statsObject[key].genotypesFreq.het = _this._freqFormatter(statsObject[key].genotypesFreq["0/1"] || statsObject[key].genotypesFreq["0|1"] || 0);
                                        statsObject[key].genotypesFreq.homalt = _this._freqFormatter(statsObject[key].genotypesFreq["1/1"] || statsObject[key].genotypesFreq["1|1"] || 0);

                                        statsObject[key].genotypesCount.homref = statsObject[key].genotypesCount["0/0"] || statsObject[key].genotypesCount["0|0"] || 0;
                                        statsObject[key].genotypesCount.het = statsObject[key].genotypesCount["0/1"] || statsObject[key].genotypesCount["0|1"] || 0;
                                        statsObject[key].genotypesCount.homalt = statsObject[key].genotypesCount["1/1"] || statsObject[key].genotypesCount["1|1"] || 0;


                                        if (key === "ALL") {
                                            statsArray.unshift({
                                                name: key,
                                                value: statsObject[key],
                                                study: study
                                            })
                                        } else {
                                            statsArray.push({
                                                name: key,
                                                value: statsObject[key],
                                                study: study
                                            })
                                        }
                                    });
                                    _variantStudies[i].stats = statsArray;
                                }
                                _this.set("variantStudies", _variantStudies);
                            }
                        })
                        .catch(function (reason) {
                            console.error(reason)
                        });
                }
            }

            fetchFacets() {
                let queryParams = {
                    sid: this.opencgaClient._config.sessionId,
                    studies: this.project.alias + ":" + this.study.alias,
                    facet: "type;genes;soAcc;studies",
                    timeout: 60000
                };
                Object.assign(queryParams, this.query);

                if (typeof queryParams.sid === "undefined" || queryParams.sid === "") {
                    delete queryParams["sid"];
                }

                let _this = this;

                // Shows loading modal
                $("#" + "MainLoadingModal").modal("show");

                setTimeout(_ => {
                    this.opencgaClient.variants().facet(queryParams, {})
                        .then(function (queryResponse) {
                            _this.facetResults = queryResponse.response[0].result[0].result.fields;
                            for (let result of _this.facetResults) {
                                let categories = [];
                                let data = [];

                                for (let countObj of result.counts) {
                                    categories.push(countObj.value);
                                    data.push(countObj.count);
                                }

                                $("#" + _this.prefix + result.name).highcharts({
                                    credits: {enabled: false},
                                    chart: {
                                        type: "column"
                                    },
                                    title: {
                                        text: result.name
                                    },
                                    xAxis: {
                                        categories: categories
                                    },
                                    yAxis: {
                                        min: 0,
                                        title: {
                                            text: 'Total number of Variants'
                                        }
                                    },
                                    tooltip: {
                                        headerFormat: '<span style="font-size:10px">{point.key}</span><table>',
                                        pointFormat: '<tr><td style="color:{series.color};padding:0">{series.name}: </td>' +
                                        '<td style="padding:0"><b>{point.y:.1f} </b></td></tr>',
                                        footerFormat: '</table>',
                                        shared: true,
                                        useHTML: true
                                    },
                                    plotOptions: {
                                        column: {
                                            pointPadding: 0.2,
                                            borderWidth: 0
                                        }
                                    },
                                    series: [{
                                        name: result.name,
                                        data: data
                                    }]
                                });
                            }

                            // Remove loading modal
                            $("#" + "MainLoadingModal").modal("hide");
                        })
                        .catch(function (e) {
                            console.log(e);
                            // Remove loading modal
                            $("#" + "MainLoadingModal").modal("hide");
                        });
                },250);

            }

            _freqFormatter(value) {
                if (value !== 0 && value !== 1) {
                    return Number(value).toFixed(5);
                }
                return value;
            }

            // This observer is needed when pie charts are rendered instead of table. This is where the data needed for high charts is prepared
            variantStudiesChanged() {
                this.$.variantStudies.render();
                if (this.variantStudies.length > 0) {
                    for (let i = 0; i < this.variantStudies.length; i++) {
                        let statsArray = this.variantStudies[i].stats;
                        for (let j = 0; j < statsArray.length; j++) {
                            let genotypeCount = statsArray[j].value.genotypesCount;
                            let data = [];
                            if (Object.keys(genotypeCount).length > 0) {
                                for (let k in genotypeCount) {
                                    data.push({
                                        name: k,
                                        y: genotypeCount[k],
                                        color: this.genotypeColor[k]
                                    });
                                }
                                let params = {title: statsArray[j].name, data: data};
                                this._updatePieChart("#" + statsArray[j].study + statsArray[j].name, params);
                            }
                        }
                    }
                }
            }

            getStudy(study) {
                let fields = study.split(':');
                return study.split(':')[fields.length - 1];
            }

            checkForGenotypes(stat) {
                let genotypesCount = stat.value.genotypesCount;
                return Object.keys(genotypesCount).length > 0;
            }

            checkForSampleData(samplesData) {
                return samplesData.length > 0;
            }

            handleCollapseAction(e) {
                let id = e.target.dataId;
                let elem = $('#' + id)[0];
                elem.hidden = !elem.hidden;
                if (elem.hidden) {
                    e.target.className = "fa fa-plus-circle";
                } else {
                    e.target.className = "fa fa-minus-circle";
                }
            }

            _changeView(e) {
                e.preventDefault(); // prevents the hash change to "#" and allows to manipulate the hash fragment as needed
                $('.variant-browser-content').hide(); // hides all content divs
                if (typeof e.target !== "undefined" && typeof e.target.dataset.view !== "undefined") {
                    $("#" + this.prefix + e.target.dataset.view).show(); // get the href and use it find which div to show
                }

                // Show the active button
                $('.variant-browser-view-buttons').removeClass("active");
                $(e.target).addClass("active");

                if (e.target.dataset.view === "Summary") {
                    this.fetchFacets();
                }

                if (e.target.dataset.view === "GenomeBrowser") {
//                    window.location.hash = "genomebrowser";
                    this._genomeBrowserActive = true;
                } else {
                    this._genomeBrowserActive = false;
                }
            }

            // This is the old Pie Chart implementation, not used right now
            _updatePieChart(_id, params) {
                // params = { title, data}
                $(_id).highcharts({
                    credits: {enabled: false},
                    chart: {
                        plotBackgroundColor: null,
                        plotBorderWidth: null,
                        plotShadow: false,
                        type: 'pie',
                        width: 220
                    },
                    title: {
                        text: params.title
                    },
                    tooltip: {
                        pointFormat: '{series.name}: <b>{point.percentage:.2f}%</b>'
                    },
                    plotOptions: {
                        pie: {
                            allowPointSelect: true,
                            cursor: 'pointer',
                            minSize: 75,
                            dataLabels: {
                                padding: 4,
                                connectorPadding: 4,
                                enabled: true,
                                format: '{point.name}: {point.percentage:.2f}%',
                                style: {
                                    color: (Highcharts.theme && Highcharts.theme.contrastTextColor) || 'black'
                                }
                            }
                        }
                    },
                    series: [{
                        name: 'Genotype Count',
                        data: params.data
                    }]
                });
            }

            triggerBeacon(e) {
                this.variantToBeacon = this.variant;
            }
        }

        customElements.define(VariantBrowserView.is, VariantBrowserView);
    </script>
</dom-module>
